src/sinks/util/udp.rs

```diff
@@ -98,197 +110,168 @@ impl UdpConnector {
-        match &mut self.state {
-            UdpServiceState::Connected(_) => Poll::Ready(Ok(())),
-            UdpServiceState::Connecting(fut) => {
-                let socket = futures::ready!(fut.poll_unpin(cx));
-                self.state = UdpServiceState::Connected(socket);
-                Poll::Ready(Ok(()))
-            }
+        loop {
+            self.state = match &mut self.state {
+                UdpServiceState::Disconnected => {
+                    let connector = self.connector.clone();
+                    UdpServiceState::Connecting(Box::pin(async move {
+                        connector.connect_backoff().await
+                    }))
+                }
+                UdpServiceState::Connecting(fut) => {
+                    let socket = ready!(fut.poll_unpin(cx));
+                    UdpServiceState::Connected(socket)
+                }
+                UdpServiceState::Connected(_) => break,
+                UdpServiceState::Sending(fut) => {
+                    let socket = match ready!(fut.poll_unpin(cx)).context(ServiceChannelRecvError) {
+                        Ok(socket) => socket,
+                        Err(error) => return Poll::Ready(Err(error)),
+                    };
+                    UdpServiceState::Connected(socket)
+                }
+            };
         }
+        Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, msg: Bytes) -> Self::Future {
-        match &mut self.state {
-            UdpServiceState::Connecting(_) => unreachable!(),
-            UdpServiceState::Connected(socket) => {
-                future::ready(udp_send(socket, &msg).context(SendError))
-            }
-        }
-    }
-}
-
-fn udp_send(socket: &mut UdpSocket, msg: &[u8]) -> Result<(), std::io::Error> {
-    let sent = socket.send(&msg)?;
-    if sent != msg.len() {
-        emit!(UdpSendIncomplete {
-            data_size: msg.len(),
-            sent,
-        });
+        let (sender, receiver) = oneshot::channel();
+
+        let mut socket =
+            match std::mem::replace(&mut self.state, UdpServiceState::Sending(receiver)) {
+                UdpServiceState::Connected(socket) => socket,
+                _ => panic!("UdpService::poll_ready should be called first"),
+            };
+
+        Box::pin(async move {
+            // TODO: Add reconnect support as TCP/Unix?
+            let result = udp_send(&mut socket, &msg).await.context(SendError);
+            let _ = sender.send(socket);
+            result
+        })
     }
-    Ok(())
 }
 
-pub struct UdpSink {
+struct UdpSink {
     connector: UdpConnector,
-    state: State,
-    span: tracing::Span,
-    backoff: ExponentialBackoff,
-}
-
-enum State {
-    Initializing,
-    Connecting(Box<dyn Future<Item = UdpSocket, Error = UdpError> + Send>),
-    Connected(UdpSocket),
-    Backoff(Box<dyn Future<Item = (), Error = ()> + Send>),
+    acker: Acker,
+    encode_event: Arc<dyn Fn(Event) -> Option<Bytes> + Send + Sync>,
```

Comment:
`Box` can be used here:

```diff
diff --git a/src/sinks/util/udp.rs b/src/sinks/util/udp.rs
index 267b084d3..a777882ab 100644
--- a/src/sinks/util/udp.rs
+++ b/src/sinks/util/udp.rs
@@ -18,7 +18,6 @@ use snafu::{ResultExt, Snafu};
 use std::{
     net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
     pin::Pin,
-    sync::Arc,
     task::{Context, Poll},
     time::Duration,
 };
@@ -210,7 +209,7 @@ impl tower::Service<Bytes> for UdpService {
 struct UdpSink {
     connector: UdpConnector,
     acker: Acker,
-    encode_event: Arc<dyn Fn(Event) -> Option<Bytes> + Send + Sync>,
+    encode_event: Box<dyn Fn(Event) -> Option<Bytes> + Send + Sync>,
 }
 
 impl UdpSink {
@@ -222,7 +221,7 @@ impl UdpSink {
         Self {
             connector,
             acker,
-            encode_event: Arc::new(encode_event),
+            encode_event: Box::new(encode_event),
         }
     }
 }
@@ -230,21 +229,15 @@ impl UdpSink {
 #[async_trait]
 impl StreamSink for UdpSink {
     async fn run(&mut self, input: BoxStream<'_, Event>) -> Result<(), ()> {
-        let encode_event = Arc::clone(&self.encode_event);
-        let mut input = input
-            // We send event empty events because we need `ack` and `emit!`.
-            .map(|event| match encode_event(event) {
-                Some(bytes) => bytes,
-                None => Bytes::new(),
-            })
-            .peekable();
+        let mut input = input.peekable();
 
         while Pin::new(&mut input).peek().await.is_some() {
             let mut socket = self.connector.connect_backoff().await;
-            while let Some(bytes) = input.next().await {
+
+            while let Some(event) = input.next().await {
+                let bytes = (self.encode_event)(event).unwrap_or_default();
                 let result = udp_send(&mut socket, &bytes).await;
                 self.acker.ack(1);
-
                 match result {
                     Ok(()) => emit!(SocketEventsSent {
                         mode: SocketMode::Udp,
@@ -255,7 +248,7 @@ impl StreamSink for UdpSink {
                         emit!(UdpSocketError { error });
                         break;
                     }
-                };
+                }
             }
         }
```